_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+there is considerable information available at parse time(e.g. literals),
+that the compiler seems to be failing to take advantage of
+then you could try a macro (or a generated function).
+But make absolutely sure to benchmark it before and after.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+compute_poly(x, coeffs) = sum(a * x^(i-1) for (i, a) in enumerate(coeffs))
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+compute_poly (generic function with 2 methods)
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+compute_poly(x, coeffs) = sum(a * x^(i-1) for (i, a) in enumerate(coeffs))
+```
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+@btime compute_poly(1, (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17))
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  49.146 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+<div class="jupyter-cell">
+
+```
+153
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+macro compute_poly(x, coeffs_tuple)
+    # a*x^i
+    Meta.isexpr(coeffs_tuple, :tupel) || ArgumentError("@compute_poly only accepts a tuple literal as second argument")
+    coeffs = coeffs_tuple.args
+    terms = map(enumerate(coeffs)) do (i, a)
+        a = coeffs[i]
+        if a isa Number && x isa Number # it is a literal compute at compile time
+            a * x ^ (i-1)
+        else
+            # an expression, so return an expression
+            esc(:($a * $x ^ $(i-1)))
+        end
+    end
+    if all(x isa Number for x in terms)
+        return sum(terms)
+    else
+        return Expr(:call, :+, terms...)
+    end
+end
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+@compute_poly (macro with 1 method)
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+@btime @compute_poly(1, (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17))
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-stream jupyter-cell">
+
+```
+
+  0.034 ns (0 allocations: 0 bytes)
+
+```
+</div>
+
+<div class="jupyter-cell">
+
+```
+153
+```
+
+</div>
+
+## Use of Dicts to hold collections of variables.
+
+`Dict`s  or hashmaps are a fantastic data structure, with expected O(1) set and get.
+However, they are not a tool for all occasions.
+I recall my undergrad data structured lecturer saying just that when we got to the last few weeks of the unit and were covering hashmap which has this great time complexity.
+As the time, having just spent most of a semester covering various lists and trees etc, I was like _"Of course not, who would do such a thing?"_.
+Now, I suspect I know.
+I think its mostly people who didn't have the fortune of a formal computer science education (e.g. most scientists and engineers) and/or never thought to benchmark just how large the constant behind that O(1) time is.
```

Comment:
Also, being from a non-CS background, I have doubts about the claim, I think it would be mostly from coming from other languages, where creating a named struct can require quite some hassle, and anonymous classes don't exist / have a heavy syntax.
These would include java, matlab, python, and there you covered mainstream languages that people come from